README.md
# adf435xcfg
This is a command line tool to allow control of the Analog Devices ADF435x devices. It
is essentially a quick and dirty port of the Python application [pyadf435x ](https://github.com/jhol/pyadf435x) to C,
and instead of using a microcontroller to form the interface between the PC and the ADF435x, it instead uses a CH341A
since that is what I had available.

Note that there are many optional settings for the ADF435x that are not exposed. If there are any that you would really like
to be able to change at the command line, please don't hesitate to contact me and i'll see what I can do, or alternatively change
the source yourself and send me a pull request and i'll merge the changes.

Hopefully this will be of use to someone else!

## CH341 <-> ADF435x Connections

|  CH341 Pin  |  ADF4350/1 Pin  |
|  ---------  |  -------------  |
|  CS0        |  LE             |
|  SCK        |  CLK            |
|  MOSI       |  DAT            |

## Binaries
A binary compiled for win64 is included. I've not built or tested this on Linux, but since it uses [libusb](https://libusb.info/), it should be possible to build this on any platform that it supports.

## Use

### Get the available options:
~~~
.\adf435xcfg.exe --help

Usage: adf435xcfg.exe <options>

  -f --frequency <freq>           Set the output frequency to <freq> Hz

  -s --sweep                       Enable sweep mode

  -l --low <freq>                  Set the sweep lower frequency to <freq> Hz

  -h --high <freq>                 Set the sweep upper frequency to <freq> Hz

  -r --resolution <freq>           Set the sweep step frequency to <freq> Hz

  -d --delay <delay>               Set the sweep mode step delay to <delay> milliseconds

  -? --help                        Display help
~~~

### Command to generate a single frequency
~~~
.\adf435xcfg.exe --freq 900000000
~~~

### Command to sweep a frequency range of 800MHz to 1GHz with 100KHz steps, delay of 1ms per step
~~~
.\adf435xcfg.exe --sweep --low 800000000 --high 1000000000 --resolution 100000 --delay 1
~~~


Changelog.txt
December 26th 2021
	First version released.

ch341.c
/****************************************************
 * 
 * This file was taken from the following project
 * 
 * https://github.com/hackpascal/ch341prog
 * 
 * All credit for it goes to its author, Weijie Gao
 * 
 ****************************************************/


#include <stdio.h>
#include <string.h>

#include "libusb.h"

#include "ch341.h"

const unsigned char BitSwapTable[256] =
{
	0x00, 0x80, 0x40, 0xC0, 0x20, 0xA0, 0x60, 0xE0,
	0x10, 0x90, 0x50, 0xD0, 0x30, 0xB0, 0x70, 0xF0,
	0x08, 0x88, 0x48, 0xC8, 0x28, 0xA8, 0x68, 0xE8,
	0x18, 0x98, 0x58, 0xD8, 0x38, 0xB8, 0x78, 0xF8,
	0x04, 0x84, 0x44, 0xC4, 0x24, 0xA4, 0x64, 0xE4,
	0x14, 0x94, 0x54, 0xD4, 0x34, 0xB4, 0x74, 0xF4,
	0x0C, 0x8C, 0x4C, 0xCC, 0x2C, 0xAC, 0x6C, 0xEC,
	0x1C, 0x9C, 0x5C, 0xDC, 0x3C, 0xBC, 0x7C, 0xFC,
	0x02, 0x82, 0x42, 0xC2, 0x22, 0xA2, 0x62, 0xE2,
	0x12, 0x92, 0x52, 0xD2, 0x32, 0xB2, 0x72, 0xF2,
	0x0A, 0x8A, 0x4A, 0xCA, 0x2A, 0xAA, 0x6A, 0xEA,
	0x1A, 0x9A, 0x5A, 0xDA, 0x3A, 0xBA, 0x7A, 0xFA,
	0x06, 0x86, 0x46, 0xC6, 0x26, 0xA6, 0x66, 0xE6,
	0x16, 0x96, 0x56, 0xD6, 0x36, 0xB6, 0x76, 0xF6,
	0x0E, 0x8E, 0x4E, 0xCE, 0x2E, 0xAE, 0x6E, 0xEE,
	0x1E, 0x9E, 0x5E, 0xDE, 0x3E, 0xBE, 0x7E, 0xFE,
	0x01, 0x81, 0x41, 0xC1, 0x21, 0xA1, 0x61, 0xE1,
	0x11, 0x91, 0x51, 0xD1, 0x31, 0xB1, 0x71, 0xF1,
	0x09, 0x89, 0x49, 0xC9, 0x29, 0xA9, 0x69, 0xE9,
	0x19, 0x99, 0x59, 0xD9, 0x39, 0xB9, 0x79, 0xF9,
	0x05, 0x85, 0x45, 0xC5, 0x25, 0xA5, 0x65, 0xE5,
	0x15, 0x95, 0x55, 0xD5, 0x35, 0xB5, 0x75, 0xF5,
	0x0D, 0x8D, 0x4D, 0xCD, 0x2D, 0xAD, 0x6D, 0xED,
	0x1D, 0x9D, 0x5D, 0xDD, 0x3D, 0xBD, 0x7D, 0xFD,
	0x03, 0x83, 0x43, 0xC3, 0x23, 0xA3, 0x63, 0xE3,
	0x13, 0x93, 0x53, 0xD3, 0x33, 0xB3, 0x73, 0xF3,
	0x0B, 0x8B, 0x4B, 0xCB, 0x2B, 0xAB, 0x6B, 0xEB,
	0x1B, 0x9B, 0x5B, 0xDB, 0x3B, 0xBB, 0x7B, 0xFB,
	0x07, 0x87, 0x47, 0xC7, 0x27, 0xA7, 0x67, 0xE7,
	0x17, 0x97, 0x57, 0xD7, 0x37, 0xB7, 0x77, 0xF7,
	0x0F, 0x8F, 0x4F, 0xCF, 0x2F, 0xAF, 0x6F, 0xEF,
	0x1F, 0x9F, 0x5F, 0xDF, 0x3F, 0xBF, 0x7F, 0xFF
};

static struct libusb_device_handle *CH341DeviceHanlde;

bool CH341DeviceInit(void)
{
	int ret;
	unsigned char desc[0x12];

	if (CH341DeviceHanlde)
		return true;

	if ((ret = libusb_init(NULL)))
	{
		fprintf(stderr, "Error: libusb_init failed: %d (%s)\n", ret, libusb_error_name(ret));
		return false;
	}

	if (!(CH341DeviceHanlde = libusb_open_device_with_vid_pid(NULL, CH341_USB_VID, CH341_USB_PID)))
	{
		fprintf(stderr, "Error: CH341 device (%04x/%04x) not found\n", CH341_USB_VID, CH341_USB_PID);
		return false;
	}

#if !defined(_MSC_VER) && !defined(MSYS) && !defined(CYGWIN) && !defined(WIN32) && !defined(MINGW) && !defined(MINGW32)
	if (libusb_kernel_driver_active(CH341DeviceHanlde, 0))
	{
		if ((ret = libusb_detach_kernel_driver(CH341DeviceHanlde, 0)))
		{
			fprintf(stderr, "Error: libusb_detach_kernel_driver failed: %d (%s)\n", ret, libusb_error_name(ret));
			goto cleanup;
		}
	}
#endif

	if ((ret = libusb_claim_interface(CH341DeviceHanlde, 0)))
	{
		printf("%s:%d\n", __FILE__, __LINE__);

		fprintf(stderr, "Error: libusb_claim_interface failed: %d (%s)\n", ret, libusb_error_name(ret));
		goto cleanup;
	}

	if (!(ret = libusb_get_descriptor(CH341DeviceHanlde, LIBUSB_DT_DEVICE, 0x00, desc, 0x12)))
	{
		printf("%s:%d\n", __FILE__, __LINE__);

		fprintf(stderr, "Warning: libusb_get_descriptor failed: %d (%s)\n", ret, libusb_error_name(ret));
	}

	printf("CH341 %d.%02d found.\n\n", desc[12], desc[13]);

	return true;

cleanup:
	printf("%s:%d\n", __FILE__, __LINE__);

	libusb_close(CH341DeviceHanlde);
	CH341DeviceHanlde = NULL;
	return false;
}

void CH341DeviceRelease(void)
{
	if (!CH341DeviceHanlde)
		return;

	libusb_release_interface(CH341DeviceHanlde, 0);
	libusb_close(CH341DeviceHanlde);
	libusb_exit(NULL);

	CH341DeviceHanlde = NULL;
}

static int CH341USBTransferPart(enum libusb_endpoint_direction dir, unsigned char *buff, unsigned int size)
{
	int ret, bytestransferred;

	if (!CH341DeviceHanlde)
		return 0;

	if ((ret = libusb_bulk_transfer(CH341DeviceHanlde, CH341_USB_BULK_ENDPOINT | dir, buff, size, &bytestransferred, CH341_USB_TIMEOUT)))
	{
		fprintf(stderr, "Error: libusb_bulk_transfer for IN_EP failed: %d (%s)\n", ret, libusb_error_name(ret));
		return -1;
	}

	return bytestransferred;
}

static bool CH341USBTransfer(enum libusb_endpoint_direction dir, unsigned char *buff, unsigned int size)
{
	int pos, bytestransferred;

	pos = 0;

	while (size)
	{
		bytestransferred = CH341USBTransferPart(dir, buff + pos, size);

		if (bytestransferred <= 0)
			return false;

		pos += bytestransferred;
		size -= bytestransferred;
	}

	return true;
}

#define CH341USBRead(buff, size) CH341USBTransfer(LIBUSB_ENDPOINT_IN, buff, size)
#define CH341USBWrite(buff, size) CH341USBTransfer(LIBUSB_ENDPOINT_OUT, buff, size)



bool CH341ChipSelect(unsigned int cs, bool enable)
{
	unsigned char pkt[4];

	static const int csio[4] = {0x36, 0x35, 0x33, 0x27};

	if (cs > 3)
	{
		fprintf(stderr, "Error: invalid CS pin %d, 0~3 are available\n", cs);
		return false;
	}

	pkt[0] = CH341_CMD_UIO_STREAM;
	if (enable)
		pkt[1] = CH341_CMD_UIO_STM_OUT | csio[cs];
	else
		pkt[1] = CH341_CMD_UIO_STM_OUT | 0x37;
	pkt[2] = CH341_CMD_UIO_STM_DIR | 0x3F;
	pkt[3] = CH341_CMD_UIO_STM_END;

	return CH341USBWrite(pkt, 4);
}

static int CH341TransferSPI(const unsigned char *in, unsigned char *out, unsigned int size)
{
	unsigned char pkt[CH341_PACKET_LENGTH];
	unsigned int i;

	if (!size)
		return 0;

	if (size > CH341_PACKET_LENGTH - 1)
		size = CH341_PACKET_LENGTH - 1;

	pkt[0] = CH341_CMD_SPI_STREAM;

	for (i = 0; i < size; i++)
		pkt[i + 1] = BitSwapTable[in[i]];

	if (!CH341USBWrite(pkt, size + 1))
	{
		fprintf(stderr, "Error: failed to transfer data to CH341\n");
		return -1;
	}

	if (!CH341USBRead(pkt, size))
	{
		fprintf(stderr, "Error: failed to transfer data from CH341\n");
		return -1;
	}

	for (i = 0; i < size; i++)
		out[i] = BitSwapTable[pkt[i]];

	return size;
}

bool CH341StreamSPI(const unsigned char *in, unsigned char *out, unsigned int size)
{
	int pos, bytestransferred;

	if (!size)
		return true;

	pos = 0;

	while (size)
	{
		bytestransferred = CH341TransferSPI(in + pos, out + pos, size);

		if (bytestransferred <= 0)
			return false;

		pos += bytestransferred;
		size -= bytestransferred;
	}

	return true;
}

bool CH341ReadSPI(unsigned char *out, unsigned int size)
{
	int pos, bytestransferred;
	unsigned char pkt[CH341_PACKET_LENGTH];

	if (!size)
		return true;

	memset(pkt, 0, sizeof (pkt));

	pos = 0;

	while (size)
	{
		bytestransferred = CH341TransferSPI(pkt, out + pos, size);

		if (bytestransferred <= 0)
			return false;

		pos += bytestransferred;
		size -= bytestransferred;
	}

	return true;
}

bool CH341WriteSPI(const unsigned char *in, unsigned int size)
{
	int pos, bytestransferred;
	unsigned char pkt[CH341_PACKET_LENGTH];

	if (!size)
		return true;

	pos = 0;

	while (size)
	{
		bytestransferred = CH341TransferSPI(in + pos, pkt, size);

		if (bytestransferred <= 0)
			return false;

		pos += bytestransferred;
		size -= bytestransferred;
	}

	return true;
}

ch341.h
/****************************************************
 * 
 * This file was taken from the following project
 * 
 * https://github.com/hackpascal/ch341prog
 * 
 * All credit for it goes to its author, Weijie Gao
 * 
 ****************************************************/

#ifndef _CH341_H_
#define _CH341_H_

#include <stdbool.h>

#define CH341_USB_VID				0x1A86
#define CH341_USB_PID				0x5512

#define CH341_USB_BULK_ENDPOINT		0x02
#define CH341_PACKET_LENGTH			0x20

#define CH341_USB_TIMEOUT			15000

#define CH341_CMD_SPI_STREAM		0xA8	//SPI command
#define CH341_CMD_UIO_STREAM		0xAB	//UIO command

#define	CH341_CMD_UIO_STM_IN		0x00	// UIO Interface In ( D0 ~ D7 )
#define	CH341_CMD_UIO_STM_DIR		0x40	// UIO interface Dir( set dir of D0~D5 )
#define	CH341_CMD_UIO_STM_OUT		0x80	// UIO Interface Output(D0~D5)
#define	CH341_CMD_UIO_STM_END		0x20	// UIO Interface End Command

bool CH341DeviceInit(void);
void CH341DeviceRelease(void);

bool CH341ChipSelect(unsigned int cs, bool enable);
bool CH341StreamSPI(const unsigned char *in, unsigned char *out, unsigned int size);
bool CH341ReadSPI(unsigned char *out, unsigned int size);
bool CH341WriteSPI(const unsigned char *in, unsigned int size);

static inline bool SPIWrite(const unsigned char *data, unsigned int size)
{
	if (!CH341ChipSelect(0, true))
		return false;
	if (!CH341WriteSPI(data, size))
		return false;
	return CH341ChipSelect(0, false);
}

static inline bool SPIRead(unsigned char *data, unsigned int size)
{
	if (!CH341ChipSelect(0, true))
		return false;
	if (!CH341ReadSPI(data, size))
		return false;
	return CH341ChipSelect(0, false);
}

static inline bool SPIWriteThenRead(const unsigned char *in, unsigned int in_size, unsigned char *out, unsigned int out_size)
{
	if (!CH341ChipSelect(0, true))
		return false;
	if (!CH341WriteSPI(in, in_size))
		return false;
	if (!CH341ReadSPI(out, out_size))
		return false;
	return CH341ChipSelect(0, false);
}

#endif /* _CH341_H_ */

common.h
/****************************************************************************
 *
 * Copyright 2021 Lee Mitchell <lee@indigopepper.com>
 * This file is part of ADF435xCFG (ADF435x Configurator)
 *
 * ADF435xCFG (ADF435x Configurator) is free software: you can redistribute it
 * and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * ADF435xCFG (ADF435x Configurator) is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ADF435xCFG (ADF435x Configurator).  If not,
 * see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************/

#ifndef COMMON_H
#define COMMON_H

/****************************************************************************/
/***        Include files                                                 ***/
/****************************************************************************/

#include <stdint.h>

/****************************************************************************/
/***        Macro Definitions                                             ***/
/****************************************************************************/

/*--------------------------------------------------------------------------*/
/*          Boolean Constants                                               */
/*--------------------------------------------------------------------------*/

#if !defined FALSE && !defined TRUE
#define TRUE            (1)   /* page 207 K+R 2nd Edition */
#define FALSE           (0)
#endif /* !defined FALSE && #if !defined TRUE */

/*--------------------------------------------------------------------------*/
/*          Null pointers                                                   */
/*--------------------------------------------------------------------------*/

/* NULL data pointer */
#if !defined NULL
#define NULL                    ( (void *) 0 )
#endif

/****************************************************************************/
/***        Type Definitions                                              ***/
/****************************************************************************/

typedef uint8_t                 bool_t;

/****************************************************************************/
/***        Exported Functions                                            ***/
/****************************************************************************/

#endif // COMMON_H

/****************************************************************************/
/***        END OF FILE                                                   ***/
/****************************************************************************/


adf435x.h
/****************************************************************************
 *
 * Copyright 2021 Lee Mitchell <lee@indigopepper.com>
 * This file is part of ADF435xCFG (ADF435x Configurator)
 *
 * ADF435xCFG (ADF435x Configurator) is free software: you can redistribute it
 * and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * ADF435xCFG (ADF435x Configurator) is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ADF435xCFG (ADF435x Configurator).  If not,
 * see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************/

#ifndef _ADF4351_H_
#define _ADF4351_H_

#include <stdbool.h>
#include <stdint.h>

typedef enum{
	E_ADF435X_VERBOSITY_LOW = 0,
	E_ADF435X_VERBOSITY_MEDIUM = 1,
	E_ADF435X_VERBOSITY_HIGH = 2,
} ADF435X_teVerbosity;

typedef enum {
    E_ADF435X_DEVICE_TYPE_ADF4350 = 0,
    E_ADF435X_DEVICE_TYPE_ADF4351 = 1,
} ADF435X_teDeviceType;

typedef enum {
    E_ADF435X_LOW_NOISE_MODE = 0,
    E_ADF435X_LOW_SPUR_MODE = 1,
} ADF435X_teLowNoiseOrLowSpurMode;

typedef enum {
    E_ADF435X_MUX_OUT_TRISTATE = 0,
    E_ADF435X_MUX_OUT_DVDD = 1,
    E_ADF435X_MUX_OUT_DGND = 2,
    E_ADF435X_MUX_OUT_R_COUNTER_OUTPUT = 3,
    E_ADF435X_MUX_OUT_N_DIVIDER_OUTPUT = 4,
    E_ADF435X_MUX_OUT_ANALOG_LOCK_DETECT = 5,
    E_ADF435X_MUX_OUT_DIGITAL_LOCK_DETECT = 6,
} ADF435X_teMuxOut;

typedef enum {
    E_ADF435X_PD_POLARITY_NEGATIVE = 0,
    E_ADF435X_PD_POLARITY_POSITIVE = 1,
} ADF435X_tePDPolarity;

typedef enum {
    E_ADF435X_BAND_SELECT_CLOCK_MODE_LOW = 0,
    E_ADF435X_BAND_SELECT_CLOCK_MODE_HIGH = 1,
} ADF435X_teBandSelectClockMode;

typedef enum {
    E_ADF435X_CLOCK_DIVIDER_MODE_OFF = 0,
    E_ADF435X_CLOCK_DIVIDER_MODE_FAST_LOCK_ENABLE = 1,
    E_ADF435X_CLOCK_DIVIDER_MODE_RESYNC_ENABLE = 2,
} ADF435X_teClockDviderMode;

typedef enum {
    E_ADF435X_FEEDBACK_SELECT_DIVIDER = 0,
    E_ADF435X_FEEDBACK_SELECT_FUNDAMENTAL = 1,
} ADF435X_teFeedbackSelect;

typedef enum {
    E_ADF435X_AUX_OUTPUT_SELECT_DIVIDED_OUTPUT = 0,
    E_ADF435X_AUX_OUTPUT_SELECT_FUNDAMENTAL = 1,
} ADF435X_teAuxOutputSelect;

typedef enum {
    E_ADF435X_LDPIN_MODE_LOW = 0,
    E_ADF435X_LDPIN_MODE_DIGITAL_LOCK_DETECT = 1,
    E_ADF435X_LDPIN_MODE_DIGITAL_LOCK_HIGH = 2,
} ADF435X_teLDPinMode;

typedef enum {
    E_ADF435X_PRESCALER_4_OVER_5 = 0,
    E_ADF435X_PRESCALER_8_OVER_9 = 1,
} ADF435X_tePrescaler;

typedef enum {
    E_ADF435X_OUTPUT_POWER_MINUS_4dBm = 0,
    E_ADF435X_OUTPUT_POWER_MINUS_1dBm = 1,
    E_ADF435X_OUTPUT_POWER_PLUS_2dBm = 2,
    E_ADF435X_OUTPUT_POWER_PLUS_5dBm = 3,
} ADF435X_teOutputPower;

typedef struct {
    ADF435X_teDeviceType eDeviceType;
    ADF435X_teFeedbackSelect eFeedbackSelect;
    ADF435X_teBandSelectClockMode eBandSelectClockMode;
    ADF435X_teLowNoiseOrLowSpurMode eLowNoiseOrLowSpurMode;
    ADF435X_teMuxOut eMuxOut;
    ADF435X_tePDPolarity ePDPolarity;
    ADF435X_teClockDviderMode eClockDivMode;
    ADF435X_teAuxOutputSelect eAuxOutputSelect;
    ADF435X_teLDPinMode eLDPinMode;
    ADF435X_tePrescaler ePrescaler;
    ADF435X_teOutputPower eOutputPower;
    ADF435X_teOutputPower eAuxOutputPower;

    uint64_t u64ReferenceFrequencyHz;
    uint64_t u64ChannelSpacingHz;

    uint32_t u32RCounter;
    uint32_t u32PhaseValue;
    uint32_t u32ClockDividerValue;

    float fChargePumpCurrent;
    float fLDP;
    float fABP;

    bool bEnableGCD;
    bool bRefDoubler;
    bool bRefDiv2;
    bool bDoubleBufR4;
    bool bPowerDown;
    bool bCPTristate;
    bool bCounterReset;
    bool bChargeCancel;
    bool bCSR;
    bool bVCOPowerDown;
    bool bMuteTillLockDetect;
    bool bAuxOutputEnable;
    bool bOutputEnable;

} ADF435x_tsOptions;

typedef struct {
    uint64_t u64Int;
    uint64_t u64Mod;
    uint64_t u64Frac;
    uint64_t u64OutputDivider;
    uint64_t u64BandSelectClockDivider;
} ADF435X_tsSettings;

typedef union {
    struct {
        uint32_t u32Register0;
        uint32_t u32Register1;
        uint32_t u32Register2;
        uint32_t u32Register3;
        uint32_t u32Register4;
        uint32_t u32Register5;
    };
    uint32_t au32[6];
} ADF435X_tuRegisters;

void ADF435x_vInit(ADF435X_teVerbosity eVerbosity);
void ADF435x_vGetOptions(ADF435x_tsOptions *psOptions);
bool ADF435x_bCalculateSettings(uint64_t u64Frequency, ADF435x_tsOptions *psOptions, ADF435X_tsSettings *psSettings);
bool ADF435x_bGenerateRegisters(ADF435x_tsOptions *psOptions, ADF435X_tsSettings *psSettings, ADF435X_tuRegisters *puRegisters);


#endif // _ADF4351_H_

main.c
/****************************************************************************
 *
 * Copyright 2021 Lee Mitchell <lee@indigopepper.com>
 * This file is part of ADF435xCFG (ADF435x Configurator)
 *
 * ADF435xCFG (ADF435x Configurator) is free software: you can redistribute it
 * and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * ADF435xCFG (ADF435x Configurator) is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ADF435xCFG (ADF435x Configurator).  If not,
 * see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************/

/****************************************************************************/
/***        Include files                                                 ***/
/****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>
#include "common.h"

#ifdef _WIN32
#include <windows.h>
#endif

#include "ch341.h"
#include "adf435x.h"

/****************************************************************************/
/***        Macro Definitions                                             ***/
/****************************************************************************/

/****************************************************************************/
/***        Type Definitions                                              ***/
/****************************************************************************/

typedef enum
{
	E_STATUS_OK,
	E_STATUS_AGAIN,
	E_STATUS_ERROR_TIMEOUT,
	E_STATUS_ERROR_WRITING,
	E_STATUS_OUT_OF_RANGE,
	E_STATUS_NULL_PARAMETER,
	E_STATUS_FAIL
} teStatus;

typedef enum{
	E_VERBOSITY_LOW,
	E_VERBOSITY_MEDIUM,
	E_VERBOSITY_HIGH
} teVerbosity;

typedef struct
{
	volatile bool_t		bExitRequest;
	uint64_t			u64Frequency;
	bool				bSweepMode;
	uint64_t			u64FreqLow;
	uint64_t			u64FreqHigh;
	uint64_t			u64FreqStep;
	teVerbosity			eVerbosity;
	int					iDelay;
} tsInstance;

/****************************************************************************/
/***        Local Function Prototypes                                     ***/
/****************************************************************************/

static void vParseCommandLineOptions(tsInstance *psInstance, int argc, char *argv[]);

#ifdef _WIN32
static BOOL WINAPI bCtrlHandler(DWORD dwCtrlType);
#endif

bool bConfigureADF435x(ADF435x_tsOptions *psOptions, uint64_t u64FrequencyHz);

/****************************************************************************/
/***        Exported Variables                                            ***/
/****************************************************************************/

/****************************************************************************/
/***        Local Variables                                               ***/
/****************************************************************************/

static tsInstance sInstance;

/****************************************************************************/
/***        Exported Functions                                            ***/
/****************************************************************************/

/****************************************************************************
 *
 * NAME: main
 *
 * DESCRIPTION:
 * Application entry point
 *
 * RETURNS:
 * int
 *
 ****************************************************************************/
int main(int argc, char *argv[])
{


	int n;

	bool_t bOk = TRUE;

	char acOut[] = {0xaa, 0x55, 0x01, 0x02, 0x03, 0x04};
	char acWord[4] = {0};

	/* Initialise application state and set some defaults */
	sInstance.bExitRequest = FALSE;
	sInstance.eVerbosity = E_VERBOSITY_MEDIUM;

	sInstance.u64Frequency = 50000000;
	sInstance.bSweepMode = false;
	sInstance.u64FreqLow = 50000000;
	sInstance.u64FreqHigh = 100000000;
	sInstance.u64FreqStep = 100000;
	sInstance.iDelay = 1;

	ADF435x_tsOptions sOptions;

	printf("+----------------------------------------------------------------------+\n" \
	"|              ADF435xCFG (ADF435x Configurator)                       |\n" \
	"| Copyright (C) 2021 Lee Mitchell <lee@indigopepper.com>               |\n" \
	"|                                                                      |\n" \
	"| This program comes with ABSOLUTELY NO WARRANTY.                      |\n" \
	"| This is free software, and you are welcome to redistribute it        |\n" \
	"| under certain conditions; See the GNU General Public License         |\n" \
	"| version 3 or later for more details. You should have received a copy |\n" \
	"| of the GNU General Public License along with this software),         |\n" \
	"| If not, see <http://www.gnu.org/licenses/>.                          |\n" \
	"+----------------------------------------------------------------------+\n\n");

    SetConsoleCtrlHandler(bCtrlHandler, TRUE);

	// Initialise options struct for ADF435x device
	ADF435x_vGetOptions(&sOptions);

	// Parse the command line options
	vParseCommandLineOptions(&sInstance, argc, argv);

	// Initialise the CH4351A UAB to SPI adapter
	if(!CH341DeviceInit())
	{
		printf("Error at line %d\n", __LINE__);
	}

	// Disable chip select
	if(!CH341ChipSelect(0, FALSE))
	{
		printf("Error at line %d\n", __LINE__);
	}

	// Initialise ADF435x functions
	ADF435x_vInit(E_ADF435X_VERBOSITY_LOW);

	if(sInstance.bSweepMode)
	{
		/* Main program loop, execute until we get a signal requesting to exit */
		while(!sInstance.bExitRequest)
		{

			for(uint64_t f = sInstance.u64FreqLow; (f <= sInstance.u64FreqHigh) && !sInstance.bExitRequest; f += sInstance.u64FreqStep)
			{
				printf("\r %d.%06dMHz    ", f / 1000000, f % 1000000);
				bConfigureADF435x(&sOptions, f);
				Sleep(sInstance.iDelay);
			}

		}

		// Switch the output off before we exit
		sOptions.bOutputEnable = false;
		bConfigureADF435x(&sOptions, 35000000);
	}
	else
	{
		bConfigureADF435x(&sOptions, sInstance.u64Frequency);
	}

	CH341DeviceRelease();

	printf("\nDone!\n");

	return EXIT_SUCCESS;
}


/****************************************************************************/
/***        Local Functions                                               ***/
/****************************************************************************/
/****************************************************************************
 *
 * NAME: vParseCommandLineOptions
 *
 * DESCRIPTION:
 * Parse command line options
 *
 * RETURNS:
 * void
 *
 ****************************************************************************/
static void vParseCommandLineOptions(tsInstance *psInstance, int argc, char *argv[])
{

	int n;
	int c;
	int index, value, from, to, port;

	static const struct option lopts[] = {
		{ "freq",			required_argument,	0, 	'f'	},

		{ "sweep",			no_argument,		0, 	's'	},

		{ "low",			required_argument,	0, 	'l'	},
		{ "high",			required_argument,	0, 	'h'	},
		{ "resolution",		required_argument,	0, 	'r'	},
		{ "delay",			required_argument,	0, 	'd'	},

        { "verbosity",     	required_argument, 	0,  'v' },

        { "help",          	required_argument, 	0,  '?' },

		{ NULL, 0, 0, 0 },
	};


	while(1)
	{

		c = getopt_long(argc, argv, "f:sl:h:r:d:v:?:h:", lopts, NULL);

		if (c == -1)
			break;

		switch(c)
		{

		case 'f':
			psInstance->u64Frequency = _atoi64(optarg);
			printf("Frequency = %d.%dMHz\n", psInstance->u64Frequency / 1000000, psInstance->u64Frequency % 1000000);
			break;

		case 's':
			psInstance->bSweepMode = true;
			printf("Sweep mode enabled\n");
			break;

		case 'l':
			psInstance->u64FreqLow = _atoi64(optarg);
			printf("Low Frequency = %d.%dMHz\n", psInstance->u64FreqLow / 1000000, psInstance->u64FreqLow % 1000000);
			break;

		case 'h':
			psInstance->u64FreqHigh = _atoi64(optarg);
			printf("High Frequency = %d.%dMHz\n", psInstance->u64FreqHigh / 1000000, psInstance->u64FreqHigh % 1000000);
			break;

		case 'r':
			psInstance->u64FreqStep = _atoi64(optarg);
			printf("Step Frequency = %d.%dMHz\n", psInstance->u64FreqStep / 1000000, psInstance->u64FreqStep % 1000000);
			break;

		case 'd':
			psInstance->iDelay = atoi(optarg);
			printf("Step Delay = %dms\n", psInstance->iDelay);
			break;

		case 'v':
			switch(atoi(optarg))
			{

			case 0:
				psInstance->eVerbosity = E_VERBOSITY_LOW;
				break;
			
			case 1:
				psInstance->eVerbosity = E_VERBOSITY_MEDIUM;
				break;
			
			case 2:
				psInstance->eVerbosity = E_VERBOSITY_HIGH;
				break;

			default:
				psInstance->eVerbosity = E_VERBOSITY_HIGH;
				break;

			}
			break;

        case '?':
		default:
			printf("\nUsage: %s <options>\n\n", argv[0]);
			puts("  -f --frequency <freq>           Set the output frequency to <freq> Hz\n\n"
				"  -s --sweep                       Enable sweep mode\n\n"
				"  -l --low <freq>                  Set the sweep lower frequency to <freq> Hz\n\n"
				"  -h --high <freq>                 Set the sweep upper frequency to <freq> Hz\n\n"

				"  -r --resolution <freq>           Set the sweep step frequency to <freq> Hz\n\n"
				"  -d --delay <delay>               Set the sweep mode step delay to <delay> milliseconds\n\n"
				// "  -v --verbosity <level>           Set verbosity level 0, 1 & 2 are valid\n\n"
				"  -? --help                        Display help\n\n"
				);
			exit(EXIT_FAILURE);
			break;
		}

	}

}


/****************************************************************************
 *
 * NAME: bCtrlHandler
 *
 * DESCRIPTION:
 * Handles Ctrl+C events
 *
 * RETURNS:
 * BOOL
 *
 ****************************************************************************/
#ifdef _WIN32
static BOOL WINAPI bCtrlHandler(DWORD dwCtrlType)
{
    switch (dwCtrlType)
    {

    case CTRL_C_EVENT:
    case CTRL_SHUTDOWN_EVENT:
    case CTRL_BREAK_EVENT:
        printf("\nExit requested\n");
        sInstance.bExitRequest = TRUE;
        return TRUE;

    default:
        return FALSE;
    }
}
#endif


bool bConfigureADF435x(ADF435x_tsOptions *psOptions, uint64_t u64FrequencyHz)
{

	char acWord[4] = {0};

	ADF435X_tsSettings sSettings;
	ADF435X_tuRegisters uRegisters;

	// Generate calculated settings, exit if there is a problem
	if(!ADF435x_bCalculateSettings(u64FrequencyHz, psOptions, &sSettings))
	{
		return false;
	}

	// Calculate the register values, exit if there is a problem
	if(!ADF435x_bGenerateRegisters(psOptions, &sSettings, &uRegisters))
	{
		return false;
	}

	// All good if here so write the 6 registers in order R5, R4, R3, R2, R1 and R0
	for(int n = 6; n > 0; n--)
	{
	
		// printf("Sending %d - %08x\n", n-1, uRegisters.au32[n-1]);

		acWord[0] = (uRegisters.au32[n-1] >> 24) & 0xff;
		acWord[1] = (uRegisters.au32[n-1] >> 16) & 0xff;
		acWord[2] = (uRegisters.au32[n-1] >> 8) & 0xff;
		acWord[3] = (uRegisters.au32[n-1] >> 0) & 0xff;

		if(!CH341ChipSelect(0, TRUE))
		{
			printf("Error at line %d\n", __LINE__);
		}

		if(CH341WriteSPI(acWord, 4) <= 0)
		{
			printf("Error at line %d\n", __LINE__);
		}

		if(!CH341ChipSelect(0, FALSE))
		{
			printf("Error at line %d\n", __LINE__);
		}

	}

	return true;
}


/****************************************************************************/
/***        END OF FILE                                                   ***/
/****************************************************************************/

